library/src/main/java/com/novoda/downloadmanager/lib/DownloadTask.java

```diff
@@ -449,6 +453,12 @@ private void checkPausedOrCanceled() throws StopRequestException {
         }
     }
 
+    private void checkClientRules() throws StopRequestException {
+        if (!downloadReadyChecker.clientAllowsToDownload(originalDownloadBatch)) {
```

Comment:
Can you please elaborate what do you mean that the batch is an invalidated?

I really don't see any difference between this and pausing or canceling a batch that is in progress. The reason that I'm saying this is because every second we query the database for that particular batch looking for a canceled or paused status.

That paused or canceled status comes from a user action -> modify a flag in the database -> the check is done on write -> an exception is thrown -> the batch is paused or canceled

The rule check is conceptually exactly the same as pausing, the only difference is that in can be paused by several reasons, not only user action.

The only difference is that instead of querying the database and looking for a paused status, we ask a client if that batch has to be paused. The rest is exactly the same.

If you could please explain why the paused mechanism is ok but the paused by client is not?

We agreed that a possible alternative was to change how we do the check. Instead of through the client callback proceed along with the paused status, querying the database.

As you've put a blocker for this task it would be really useful if you could explain your alternative solution in order to help to unblock it :)